core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -549,8 +558,17 @@ public void writeTo(StreamOutput out) throws IOException {
     public String toString() {
         String sSource = "_na_";
         try {
-            sSource = XContentHelper.convertToJson(source, false);
-        } catch (Exception e) {
+            if (source.length() > MAX_SOURCE_LENGTH_IN_TOSTRING) {
+                // We might slice in the middle of a UTF-8 character here, but BytesRef.utf8ToString() will read more bytes if needed.
+                // There is no risk of ArrayIndexOutOfBoundsException either as source is assumed to be properly UTF-8 encoded.
+                sSource = source.slice(0, MAX_SOURCE_LENGTH_IN_TOSTRING).utf8ToString() + "..." +
```

Comment:
@imotov
The problem with throwing an exception from modified `OutputStream` is that we might stop in the middle of UTF-8 sequence. And then we'd get either `ArrayIndexOutOfBounds` in `UnicodeUtil.UTF8toUTF16(...)` or some unspecified behavior in `new String(data, "UTF-8")`.

There is a workaround: collect `fragmentLength + 3` bytes, but only use `fragmentLength` bytes for decoding with `UnicodeUtil` (which would then read more bytes if needed). But this would be quite an ugly hack.

Another option is to print CBOR, SMILE, etc. bytes in hex, a-la "03 1A B7 ..." instead of converting them to JSON. Or just omit any non-JSON source altogether. This would require guessing the content type though.

Maybe we should then just skip any large source? And only print the message like "Source is too big - 5kb". Having the excerpt however might be quite useful for debugging.